pkg/mcp/mcp.go

```diff
@@ -250,6 +263,47 @@ func mergePrompts(base, override []api.ServerPrompt) []api.ServerPrompt {
 	return result
 }
 
+// metricsMiddleware returns a metrics middleware with access to the server's metrics system
+func (s *Server) metricsMiddleware() func(mcp.MethodHandler) mcp.MethodHandler {
+	return func(next mcp.MethodHandler) mcp.MethodHandler {
+		return func(ctx context.Context, method string, req mcp.Request) (mcp.Result, error) {
+			start := time.Now()
+			result, err := next(ctx, method, req)
+			duration := time.Since(start)
+
+			toolName := method
+			if method == "tools/call" {
+				if params, ok := req.GetParams().(*mcp.CallToolParamsRaw); ok {
+					if toolReq, _ := GoSdkToolCallParamsToToolCallRequest(params); toolReq != nil {
+						toolName = toolReq.Name
+					}
+				}
+			}
+
+			// Record to all collectors
+			s.metrics.RecordToolCall(ctx, toolName, duration, err)
+
+			return result, err
+		}
+	}
+}
```

Comment:
@nader-ziada I believe that in here (or at least, somewhere) we should be extracting traceparent/tracestate from `req.Params.Meta` (see the current proposed semconv for otel here: https://github.com/open-telemetry/semantic-conventions/pull/2083/changes)